idea/src/org/jetbrains/kotlin/idea/inspections/MainFunctionReturnUnitInspection.kt

```diff
@@ -10,39 +10,54 @@ import com.intellij.codeInspection.ProblemDescriptor
 import com.intellij.codeInspection.ProblemHighlightType
 import com.intellij.codeInspection.ProblemsHolder
 import com.intellij.openapi.project.Project
-import com.intellij.openapi.util.TextRange
 import com.intellij.psi.PsiElementVisitor
 import org.jetbrains.kotlin.builtins.KotlinBuiltIns
 import org.jetbrains.kotlin.descriptors.FunctionDescriptor
 import org.jetbrains.kotlin.idea.MainFunctionDetector
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
 import org.jetbrains.kotlin.idea.core.setType
 import org.jetbrains.kotlin.idea.search.usagesSearch.descriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtAnnotationEntry
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.namedFunctionVisitor
-import org.jetbrains.kotlin.psi.psiUtil.endOffset
 import org.jetbrains.kotlin.psi.psiUtil.getNonStrictParentOfType
-import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
-import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
 
 class MainFunctionReturnUnitInspection : AbstractKotlinInspection() {
     override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
         return namedFunctionVisitor(fun(function: KtNamedFunction) {
-            if (function.name != "main") return
+            val isMain = function.name == "main"
+            val testAnnotations = function.annotationEntries.filter { "Test" in it.text }
+            if (!isMain && testAnnotations.isEmpty()) return
 
             val descriptor = function.descriptor as? FunctionDescriptor ?: return
-            if (!MainFunctionDetector.isMain(descriptor, checkReturnType = false)) return
-            if (MainFunctionDetector.isMainReturnType(descriptor)) return
+            if (isMain) {
+                if (!MainFunctionDetector.isMain(descriptor, checkReturnType = false)) return
+            } else {
+                val junitTestFqNames = listOf(FqName("org.junit.Test"), FqName("org.junit.jupiter.api.Test"))
+                if (testAnnotations.none { it.fqName() in junitTestFqNames }) return
+            }
+            if (descriptor.returnType?.let { KotlinBuiltIns.isUnit(it) } == true) return
 
             holder.registerProblem(
                 function.nameIdentifier ?: function,
-                "main should return Unit",
+                "${if (isMain) "main" else "test"} should return Unit",
```

Comment:
I'd suggest to report "JUnit test should return Unit" or "Test function should return Unit" or something like. "test should return Unit" looks not very pretty